### PR TITLE
add an update step to replace digicert with onecert

### PR DESCRIPTION
**Pull request by yithian**
### Which issue this PR addresses:

Fixes ARO-10309

### What this PR does / why we need it:

This allows upgrades to 4.16 to proceed. These upgrades were being blocked by the SHA-1 signing algorithm used by DigiCert, which is incompatible with 4.16

### Test plan for issue:

I'm not sure how to test this. It will need to be deployed somewhere with access to an affected cluster

### Is there any documentation that needs to be updated for this PR?

No, it's mitigation for an incident

### How do you know this will function as expected in production? 

The new step is limited to affected clusters only

### CODE REVIEW

File: pkg/cluster/replacedigicert.go
Comment (tsatam) [reply pkg/cluster/replacedigicert.go:17]: I'd even go as far as to suggest this function should be called something like "ensureCertificateIssuer" or something generic, to go in line with your suggestion above to abstract the specifics around what issuer is being managed here.

Comment (yithian) [reply pkg/cluster/replacedigicert.go:17]: Due to [how `steps.actionFunction` is defined](https://github.com/Azure/ARO-RP/blob/master/pkg/util/steps/action.go#L16), I don't think the high-level step to convert from DigiCert to OneCert can be a generic `ensureCertificateIssuer`-style function with parameters

I can extract the logic for that into a generic util function though

**Review by tsatam — COMMENTED** [review]
Looks good, will approve after the PR is tested in an applicable environment to ensure it behaves as expected.

**Review by fahlmant — COMMENTED** [review]
LGTM! Thanks!

**Comment by yithian** [comment]
I flipped the cert policy issuer on my test cluster back to DigiCert after pushing my latest RP changes to canary. After running a PUCM, the (DigiCert-issued) certificate was replaced with one from OneCert

**Review by cadenmarchese — COMMENTED** [review]
There are a number of utility functions related to managed domains under https://github.com/Azure/ARO-RP/blob/master/pkg/util/dns/dns.go - can we use any of them in this PR?

File: pkg/cluster/correct_cert_issuer.go
Comment (kimorris27) [inline pkg/cluster/correct_cert_issuer.go:20]: Nit (and tangentially related to your changes): we could turn this into a "constant" by making a util method on the cluster manager type that returns the API cert's Secret name. This would ensure that it's correct across the different files that need to know and use the Secret name. Same applies to the ingress cert's Secret name.

File: pkg/cluster/correct_cert_issuer.go
Comment (kimorris27) [inline pkg/cluster/correct_cert_issuer.go:22]: Nit: the OneCert issuer name could be a constant since we use it in more than one place.

File: pkg/cluster/correct_cert_issuer.go
Comment (kimorris27) [inline pkg/cluster/correct_cert_issuer.go:50]: Nit: I'd be pretty surprised if we found any nil pointers in a non-error result of `GetCertificate`, but I'd still check them all to be super safe.

**Comment by fahlmant** [comment]
@yithian What's left to do here before merge?

**Comment by yithian** [comment]
> @yithian What's left to do here before merge?

@fahlmant I need to write up the util function (and related tests) that Kipp described [here](https://github.com/Azure/ARO-RP/pull/3867#discussion_r1802066128)

I have fixes for the other comments staged up locally but I've been fully dedicated to PUCMs for the past several days

**Comment by fahlmant** [comment]
This was supposed to go into the PUMC. I'm concerned this is getting kicked down the road. Is it possible to get this updated this week?

**Comment by yithian** [comment]
Yes, this should be done this week

**Review by kimorris27 — APPROVED** [review]
LGTM with the context that it's been tested successfully 🚀

**Review by fahlmant — DISMISSED** [review]
lgtm again :)

**Review by fahlmant — APPROVED** [review]

**Review by kimorris27 — APPROVED** [review]

**Review by cadenmarchese — APPROVED** [review]

**Pull request merged**
